src/App.js

```diff
@@ -24,62 +24,87 @@ import getToken from './lib/tokens';
 
 function App() {
   const [listId, setListId] = useState(null);
-  const [userToken, setUserToken] = useState('');
 
-  function saveToken(e) {
-    e.preventDefault();
+  function createList() {
     const token = getToken();
-    localStorage.setItem('token', token);
-    setUserToken(token);
+
+    return db
+      .collection('lists')
+      .add({ token: token })
+      .then((results) => {
+        const { id } = results;
+        if (id === undefined) throw new Error('Failed to create list.');
+        return id;
+      })
+      .then((newListId) => {
+        // list successfully created in db, so update listId state and save token to localStorage
+        setListId(newListId);
+        localStorage.setItem('token', token);
+        return true;
+      });
+  }
+
+  // returns true if token matches a list in db, returns false if no list, throws error on connection problem
+  function isTokenValid(token) {
+    return db
+      .collection('lists')
+      .where('token', '==', token)
+      .get()
+      .then((querySnapshot) => {
+        // if there are results and an id property exists
+        if (!querySnapshot.empty && 'id' in querySnapshot.docs[0]) {
+          setListId(querySnapshot.docs[0].id);
+          return true;
+
+          // check metadata.fromCache to distinguish between no results (invalid token) and a connection issue
+        } else if (!querySnapshot.metadata.fromCache) {
+          return false;
+        } else {
+          throw new Error('Connection problem');
+        }
+      });
+  }
+
+  function joinList(shareToken) {
+    return isTokenValid(shareToken).then((listExists) => {
+      if (listExists) {
+        localStorage.setItem('token', shareToken);
+        return true;
+      } else throw new Error('Invalid token');
+    });
   }
 
   // on component mounting, look for token in local storage and use it to retrieve the list id
   useEffect(() => {
     const token = localStorage.getItem('token');
 
     if (token) {
-      setUserToken(token);
-
-      // find the list in Firestore associated with the stored token from local storage
-      db.collection('lists')
-        .where('token', '==', token)
-        .get()
-        .then((querySnapshot) => {
-          // if there are results and an id property exists
-          if (!querySnapshot.empty && 'id' in querySnapshot.docs[0]) {
-            setListId(querySnapshot.docs[0].id); // save the list id for later
-          }
+      isTokenValid(token)
+        .then((listExists) => {
+          if (listExists !== true) localStorage.removeItem('token');
```

Comment:
When our component mounts, we are setting token to what is in local storage. Then we are checking `isTokenValid` against the database. If the `listExists` does not exist in the db, and thus is not equal to true, aka it is false, we are removing the token from local storage correct?

If this happened, we're assuming the user has an invalid token in local storage for some reason, so we are removing it, and the user would see the Home screen where they can create a new list/token or join an existing list/token.

I want to make sure my understanding is correct!

Reply:
Yes, we check localStorage for an existing token when the app first loads. Our function isTokenValid checks the database for the token, and also accounts for the possibility that an empty result might be because there is no list for that token (therefore it is invalid, the function returns false, and we delete the bad token) or might be because the person is having a connection issue, in which case we leave things alone and just print an error to the console. Line 47 comment gives a quick overview of the isTokenValid function and possible return values and what they indicate.